Fix permission request loop for camera access when permissions are denied permanently

maestrovsmaster opened the pull request:
@lexlog  Resolved an issue where the app entered a loop in the camera permission request process when permissions were denied, especially when the user selected "Never ask again" or permanently blocked permission requests. Improved the permission request logic by adding checks to avoid unnecessary repeated requests. If permissions are disabled, the app now redirects the user to the settings. If the user does not grant access in settings, the app gracefully navigates back to the main screen.

irinaBerendeeva87 reviewed (changes requested):
### Checked:
❌ there are  no transitions  when permissions are disabled
just white screen and error in logcat
2025-01-27 10:38:06.249  1158-3459  WindowManager           system_server                        E  win=Window{813df06 u0 io.forus.me.dev/io.forus.me.android.presentation.view.screens.dashboard.DashboardActivity} destroySurfaces: appStopped=true win.mWindowRemovalAllowed=false win.mRemoveOnExit=false win.mViewVisibility=8 caller=com.android.server.wm.AppWindowToken.destroySurfaces:1200 com.android.server.wm.AppWindowToken.destroySurfaces:1181 com.android.server.wm.AppWindowToken.notifyAppStopped:1236 com.android.server.wm.ActivityRecord.activityStoppedLocked:2586 com.android.server.wm.ActivityTaskManagerService.activityStopped:2385 android.app.IActivityTaskManager$Stub.onTransact:2180 android.os.Binder.execTransactInternal:1021 

✅ with allow permission
❌ when change perm from allow to denay/"Never ask again" 
user sees a error screen with retry but retry lead to   <-- 401 Unauthorized https://dev.api.forus.io/api/v1/identity/record-categories  and   io.forus.me.android.data.exception.RetrofitException: 401 Unauthorized


--------------------------
 
 **Not related to this PR**
 
 after restart the app I got an error in console  that could lead to NullPointer Exception
 <-- 422 Unprocessable Content https://dev.api.forus.io/api/v1/platform/firestore-tokens (367ms)
2025-01-27 11:16:32.277 21078-21172 OkHttp                  io.forus.me.dev                      D  <-- 201 Created https://dev.api.forus.io/api/v1/platform/devices/register-push

maestrovsmaster commented:
@lexlog 
Fix: Ensure proper app initialization after permission changes

Reason:  
Previously, when system permissions (e.g., CAMERA) were modified, Android restarted the DashboardActivity without reinitializing critical components like SharedPreferences and the database. This led to app crashes and unexpected behavior.

Solution:  
- Moved global initialization ( preferences) to the Application class to ensure they persist across activity restarts.  
- Added an unlock database in DashboardActivity to verify initialization before proceeding. If initialization fails, the user is redirected to MainActivity.  
- Ensured that essential components are properly reloaded when the app resumes after permission changes.

Now, the app remains stable even after system permission modifications.

RobinMeles commented:
✅ Redirected users to the app settings if permissions are blocked, when opening QR-scanner
❌ Both the main voucher screen and the profile screen are blocked and unauthorized when camera permission is turned off (it should always open these screens even if permission for camera is disabled, should only block QR-code screen)
![image](https://github.com/user-attachments/assets/075e8ec2-1405-4b2d-b633-d7796fd8b29f)
![image](https://github.com/user-attachments/assets/bb6b3ee1-2b24-44df-ad82-f1feea59c60f)
io.forus.me.android.data.exception.RetrofitException: 401 Unauthorized
at io.forus.me.android.data.net.RxErrorHandlingCallAdapterFactory$RxCallAdapterWrapper.asRetrofitException(RxErrorHandlingCallAdapterFactory.java:71)
at io.forus.me.android.data.net.RxErrorHandlingCallAdapterFactory$RxCallAdapterWrapper.-$$Nest$masRetrofitException(Unknown Source:0)
at io.forus.me.android.data.net.RxErrorHandlingCallAdapterFactory$RxCallAdapterWrapper$1.apply(RxErrorHandlingCallAdapterFactory.java:60)
at io.forus.me.android.data.net.RxErrorHandlingCallAdapterFactory$RxCallAdapterWrapper$1.apply(RxErrorHandlingCallAdapterFactory.java:57)
at io.reactivex.internal.operators.observable.ObservableOnErrorNext$OnErrorNextObserver.onError(ObservableOnErrorNext.java:91)
at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:59)
at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:38)
at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:48)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:35)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableOnErrorNext.subscribeActual(ObservableOnErrorNext.java:38)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableFlatMap.subscribeActual(ObservableFlatMap.java:55)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableSingleSingle.subscribeActual(ObservableSingleSingle.java:35)
at io.reactivex.Single.subscribe(Single.java:3666)
at io.reactivex.internal.operators.single.SingleZipArray.subscribeActual(SingleZipArray.java:63)
at io.reactivex.Single.subscribe(Single.java:3666)
at io.reactivex.internal.operators.single.SingleToObservable.subscribeActual(SingleToObservable.java:35)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableZip$ZipCoordinator.subscribe(ObservableZip.java:110)
at io.reactivex.internal.operators.observable.ObservableZip.subscribeActual(ObservableZip.java:72)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableSingleSingle.subscribeActual(ObservableSingleSingle.java:35)
at io.reactivex.Single.subscribe(Single.java:3666)
at io.reactivex.internal.operators.single.SingleZipArray.subscribeActual(SingleZipArray.java:63)
at io.reactivex.Single.subscribe(Single.java:3666)
at io.reactivex.internal.operators.single.SingleToObservable.subscribeActual(SingleToObservable.java:35)
at io.reactivex.Observable.subscribe(Observable.java:12284)
at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:608)
at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)

maestrovsmaster commented:
@RobinMeles   This error occurs when there are changes in the configuration (specifically, changes in application permissions), causing DashboardActivity to be recreated. As a result, the database session is closed, and I reopen it. This should restore the session, but only if the user does not have a PIN code set.

If the user has a PIN code enabled, opening the database from the current activity is not possible. To handle this case, I added a check in the current DashboardActivity to verify whether the database is activated. If it is not, the user is redirected to the initial screen to enter their PIN code, after which the database can be activated.

If the PIN code is disabled, the session restoration happens automatically for the user.
